avro-serde/src/main/java/io/confluent/kafka/streams/serdes/GenericAvroSerde.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2017 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.confluent.kafka.streams.serdes;
+
+import org.apache.avro.generic.GenericRecord;
+import org.apache.kafka.common.serialization.Deserializer;
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.common.serialization.Serdes;
+import org.apache.kafka.common.serialization.Serializer;
+
+import java.util.Map;
+
+import io.confluent.kafka.schemaregistry.client.SchemaRegistryClient;
+
+/**
+ * A serde (serializer/deserializer) for Apache Kafka's Streams API that can be used for for reading
+ * and/or writing data in "generic Avro" format.
+ *
+ * <p>Usage example for configuring this serde as a Kafka Streams application's default serde for
+ * both record keys and record values:
+ *
+ * <p><pre>{@code
+ * Properties streamsConfiguration = new Properties();
+ * streamsConfiguration.put(StreamsConfig.KEY_SERDE_CLASS_CONFIG, GenericAvroSerde.class);
+ * streamsConfiguration.put(StreamsConfig.VALUE_SERDE_CLASS_CONFIG, GenericAvroSerde.class);
+ * streamsConfiguration.put(
+ *     AbstractKafkaAvroSerDeConfig.SCHEMA_REGISTRY_URL_CONFIG,
+ *     "http://confluent-schema-registry-server:8081/");
+ * }</pre>
+ *
+ * <p>Usage example for explicitly overriding the application's default serdes (whatever they were
+ * configured to) so that only specific operations such as {@code KStream#to()} use this serde:
+ *
+ * <p><pre>{@code
+ * Serde<GenericRecord> genericAvroSerde = new GenericAvroSerde();
+ * boolean isKeySerde = false;
+ * genericAvroSerde.configure(
+ *     Collections.singletonMap(
+ *         AbstractKafkaAvroSerDeConfig.SCHEMA_REGISTRY_URL_CONFIG,
+ *         "http://confluent-schema-registry-server:8081/"),
+ *     isKeySerde);
+ * KStream<String, GenericRecord> stream = ...;
+ * stream.to(Serdes.String(), genericAvroSerde, "my-output-topic");
+ * }</pre>
+ */
+public class GenericAvroSerde implements Serde<GenericRecord> {
+
+  private final Serde<GenericRecord> inner;
+
+  public GenericAvroSerde() {
+    inner = Serdes.serdeFrom(new GenericAvroSerializer(), new GenericAvroDeserializer());
+  }
+
+  /**
+   * For testing purposes only.
+   */
+  GenericAvroSerde(final SchemaRegistryClient client) {
```

Comment:
To prevent multiple copies of the schema cache we are going to need to reuse the SchemaRegistryClient across Serdes. This is one of the weirdnesses I introduced into the original pull I put together. What are our plans for this?